### PR TITLE
[GEOS-10892] Allow configuring custom links for OGC API collections and single collection resources

### PR DESCRIPTION
**Pull request by aaime**
[GEOS-10892](https://osgeo-org.atlassian.net/browse/GEOS-10892)

Generic link support. Helps with INSPIRE compliance and allows to add all sorts of links for whatever other spec/convention/requirement out there.

# Checklist

- [ ] I have read the [contribution guidelines](https://github.com/geoserver/geoserver/blob/main/CONTRIBUTING.md).
- [ ] I have sent a [Contribution Licence Agreement](https://docs.geoserver.org/latest/en/developer/policies/committing.html) (not required for small changes, e.g., fixing typos in documentation).
- [ ] First PR targets the `main` branch (backports managed later; ignore for branch specific issues).
- [ ] All the build checks are green ([see automated QA checks](https://docs.geoserver.org/latest/en/developer/qa-guide/index.html)).

For core and extension modules:

- [ ] New unit tests have been added covering the changes.
- [ ] [Documentation](https://github.com/geoserver/geoserver/tree/main/doc/en/user/source) has been updated (if change is visible to end users).
- [ ] The [REST API docs](https://github.com/geoserver/geoserver/tree/main/doc/en/api/1.0.0) have been updated (when changing configuration objects or the REST controllers).
- [ ] There is an issue in the [GeoServer Jira](https://osgeo-org.atlassian.net/browse/GEOS/summary) (except for changes that do not affect administrators or end users in any way).
- [ ] Commit message(s) must be in the form ``[GEOS-XYZWV] Title of the Jira ticket``.
- [ ] Bug fixes and small new features are presented as a single commit.
- [ ] Each commit has a single objective (if there are multiple commits, each has a separate JIRA ticket describing its goal).

### CODE REVIEW

**Comment by aaime** [comment]
Hum... I can't replicate the build failure... the LinkInfo class is there, the dependency is too, not sure what's going on here. Compiles fine locally, I don't have outstanding files... can you replicate it on your end?

**Comment by turingtestfail** [comment]
I am able to replicate it with CLI Maven.  If I specifically "Rebuild
Module gs-web-ogcapi" in IntelliJ the error shows up there even though the
LinkInfo class is present and properly indexed by the IDE.

File: src/community/pom.xml [hunk 1]
```diff
@@ -16,6 +16,9 @@
   <artifactId>community</artifactId>
   <packaging>pom</packaging>
   <name>Community Space</name>
+  <modules>
+    <module>ogcapi/web-ogcapi</module>
+  </modules>
```
Comment (aaime) [inline src/community/pom.xml:21]: This is the issue... IntelliJ trying to be helpful and adding the new module in the wrong pom...

File: doc/en/user/source/community/ogc-api/configure.rst [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+Link editor column description:
+
+* **rel**: the link relation type, as per the OGC API - Features specification
+* **Mime type**: the mime type for the resource found following the link
+* **URL**: the link URL
+* **Title**: the link title (optional)
+* **Service**: the service for which the link is valid (optional, defaults to all) 
+
+
+Common links relationships that could be added for the ``collections`` resource are:
```
Comment (turingtestfail) [inline doc/en/user/source/community/ogc-api/configure.rst:33]: Even though, they refer to the same concept, I think this should be "link relations"

File: src/community/ogcapi/ogcapi-core/src/main/java/org/geoserver/ogcapi/LinkInfoCallback.java [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+package org.geoserver.ogcapi;
+
+import org.geoserver.catalog.PublishedInfo;
+import org.geoserver.catalog.ResourceInfo;
+import org.geoserver.config.GeoServer;
+import org.geoserver.config.SettingsInfo;
+import org.geoserver.ows.Request;
+import org.geoserver.platform.Service;
+import org.springframework.stereotype.Component;
+
```
Comment (turingtestfail) [inline src/community/ogcapi/ogcapi-core/src/main/java/org/geoserver/ogcapi/LinkInfoCallback.java:14]: Class javadoc?

File: doc/en/user/source/community/ogc-api/configure.rst (hunk 2 above)
Comment (aaime) [reply doc/en/user/source/community/ogc-api/configure.rst:33]: Quoting from OGC tutorials:

---
The server identifies the relationship between a resource and other linked resources through a link relation type, represented by the attribute ‘rel’.
---

So yep, "relation" it is, indeed!

File: src/community/ogcapi/ogcapi-core/src/main/java/org/geoserver/ogcapi/OGCAPIXStreamPersisterInitializer.java [hunk 4]
```diff
@@ -0,0 +1,24 @@
...
+ */
+package org.geoserver.ogcapi;
+
+import com.thoughtworks.xstream.XStream;
+import java.util.List;
+import org.geoserver.config.util.XStreamPersister;
+import org.geoserver.config.util.XStreamPersisterInitializer;
+import org.geoserver.ogcapi.impl.LinkInfoImpl;
+import org.springframework.stereotype.Component;
+
```
Comment (turingtestfail) [inline src/community/ogcapi/ogcapi-core/src/main/java/org/geoserver/ogcapi/OGCAPIXStreamPersisterInitializer.java:13]: Class javadoc?

File: src/community/ogcapi/web-ogcapi/src/main/java/org/geoserver/web/ogcapi/LinkInfoEditor.java [hunk 5]
```diff
@@ -0,0 +1,187 @@
...
+
+    private static final LoadableDetachableModel<List<Property<LinkInfo>>> propertiesModel =
+            new LoadableDetachableModel<>() {
+                @Override
+                protected List<Property<LinkInfo>> load() {
+                    return Arrays.asList(REL, TYPE, HREF, TITLE, SERVICE, REMOVE);
+                }
+            };
+
+    public LinkInfoEditor(String id, IModel<List<LinkInfo>> links, Component parent) {
```
Comment (turingtestfail) [inline src/community/ogcapi/web-ogcapi/src/main/java/org/geoserver/web/ogcapi/LinkInfoEditor.java:68]: Javadoc?

File: src/community/ogcapi/web-ogcapi/src/main/java/org/geoserver/web/ogcapi/OGCAPILayerConfig.java [hunk 6]
```diff
@@ -0,0 +1,32 @@
...
+import java.util.List;
+import org.apache.wicket.model.IModel;
+import org.apache.wicket.model.PropertyModel;
+import org.geoserver.catalog.LayerInfo;
+import org.geoserver.catalog.MetadataMap;
+import org.geoserver.ogcapi.LinkInfo;
+import org.geoserver.web.publish.PublishedConfigurationPanel;
+import org.geoserver.web.util.MetadataMapModel;
+
+public class OGCAPILayerConfig extends PublishedConfigurationPanel<LayerInfo> {
```
Comment (turingtestfail) [inline src/community/ogcapi/web-ogcapi/src/main/java/org/geoserver/web/ogcapi/OGCAPILayerConfig.java:16]: Class javadoc?

**Comment by aaime** [comment]
Feedback hopefully addressed

**Review by turingtestfail — APPROVED** [review]

**Pull request merged**